storage/innobase/log/log0recv.cc

```diff
@@ -2569,19 +2657,78 @@ recv_sys_t::parse_mtr_result recv_sys_t::parse(source &l, bool if_exists)
                       }
                     });
 
-  if (crc != (l + 1).read4())
-    return GOT_EOF;
-
-  l+= 5;
-  ut_d(const source el{l});
-  lsn+= l - begin;
+  parse_mtr_result r;
+  const size_t s{l - begin};
+  constexpr size_t tail_size{format == log_t::FORMAT_10_8 ? 1 + 4 : 1 + 8 + 4};
+  l+= tail_size;
   offset= l.ptr - log_sys.buf;
   if (!l.is_mmap());
   else if (offset == log_sys.file_size)
     offset= log_sys.START_OFFSET;
   else
     ut_ad(offset < log_sys.file_size);
 
+  l= begin;
+
+  if (format == log_t::FORMAT_ENC_11)
+  {
+    decrypt_buffer_offset= begin.offset();
+    const byte *ptr{static_cast<byte*>(begin.memcpy(decrypt_buffer, s + 9))};
+    log_decrypt_mtr(const_cast<byte*>(ptr), ptr + s);
+    recv_clear c{ptr};
+    r= parse_tail<recv_clear,storing,format>(c, if_exists, s + tail_size);
+    ut_ad(c.ptr >= ptr);
+    if (storing == YES && r == GOT_RESTART)
+    {
+      l+= c.ptr - ptr;
+      goto restart;
+    }
+  }
+  else
+  {
+    r= parse_tail<source,storing,format>(l, if_exists, s + tail_size);
+    if (storing == YES && r == GOT_RESTART)
+      goto restart;
+  }
+  ut_ad(r != GOT_RESTART);
+  return r;
+}
+
+ATTRIBUTE_COLD void recv_sys_t::rewind_prepare(lsn_t start_lsn) noexcept
+{
+  if (start_lsn > log_sys.get_flushed_lsn(std::memory_order_relaxed))
+    log_sys.set_recovered_lsn(start_lsn);
+  lsn= start_lsn;
+}
+
+ATTRIBUTE_COLD recv_sys_t::parse_mtr_result recv_sys_t::rewind_apply() noexcept
+{
+  apply(false);
+  return is_corrupt_fs() ? GOT_EOF : GOT_RESTART;
+}
+
+/** Parse and register one mini-transaction.
+@tparam source    type of log data source
+@tparam storing   whether to store the records
+@tparam format    log record format
+@param  begin     start of the mini-transaction
+@param  if_exists if storing==YES: whether to check if the tablespace exists
+@param  size      size of the mini-transaction */
+template<typename source,recv_sys_t::store storing,uint32_t format>
+inline recv_sys_t::parse_mtr_result
+recv_sys_t::parse_tail(source &begin, const bool if_exists, size_t size)
+  noexcept
+{
+  byte *const iv= format != log_t::FORMAT_ENC_10_8 ? nullptr :
+    static_cast<byte*>(alloca(MY_AES_BLOCK_SIZE));
```

Comment:
Now that we will allocate `recv_sys.decrypt_buffer` from the heap, we could use a part of that one also for the old `FORMAT_ENC_10_8` initialization vector. I will need to check the generated code whether this is worth the complication.